src/misc/reachability.c

```diff
@@ -0,0 +1,122 @@
+/* -*- mode: C -*-  */
+/* vim:set ts=4 sw=4 sts=4 et: */
+/*
+   IGraph library.
+   Copyright (C) 2024  The igraph development team <igraph@igraph.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "igraph_adjlist.h"
+#include "igraph_community.h"
+#include "igraph_components.h"
+#include "igraph_interface.h"
+#include "igraph_reachability.h"
+
+void _print_vector_int(const igraph_vector_int_t *v) {
+    igraph_integer_t i, n = igraph_vector_int_size(v);
+    printf("(");
+    for (i=0; i < n; i++) {
+        printf(" %" IGRAPH_PRId, VECTOR(*v)[i]);
+    }
+    printf(" )\n");
+}
+
+igraph_integer_t igraph_bitset_popcount(igraph_vector_int_t *bitset)
+{
+    igraph_integer_t i, n, count = 0;
+    n = igraph_vector_int_size(bitset);
+    for (i = 0; i < n; ++i)
+    {
+        // TODO: Check if __cpuid claims the operation is supported by CPU
+        // https://learn.microsoft.com/en-us/cpp/intrinsics/popcnt16-popcnt-popcnt64?view=msvc-170
+        #ifdef _MSC_VER
+        count += __popcnt64(VECTOR(*bitset)[i]);
```

Comment:
Ultimately we should test for `__popcnt64` and `__builtin_popcountll` with specific tests from CMake at the configuration stage and we should provide `HAVE_POPCNT64` and `HAVE_BUILTIN_POPCOUNTLL` macros in `config.h`.

Also, I was wondeirng whether this takes into account that igraph integers can be configured as 32-bit or 64-bit (these are the two possible values for `IGRAPH_INTEGER_SIZE`).

Reply:
No, it currently assumes 64-bit but I will change that.

I will add all those checks and fallback to implementation that doesn't require the popcount instruction when unavailable.
ARM and maybe M2 may also need specialization, not sure as I have not developed for those.

Reply:
The CPU checks may need to happen at runtime. The binary may be built on a different machine than runs it, right?

Reply:
The binary is built for a certain architecture and instruction set so these do not have to be runtime checks. It _may_ happen that you build the binary for an instruction set and run it on a machine that has a _wider_ instruction set (like, you can build for `armv6` and then run it on a machine that supports `armv7`), but this is rare. And I don't think it's even possible to perform this check in runtime.

Reply:
It is possible to perform the checks at runtime using `__cpuid` or `__builtin_cpu_supports` for example.
Then using a dispatch function call to set the function pointer to the correct function on its first run you can ensure that the correct function is called depending on what features are available. This way you do not waste CPU instructions on determining the correct implementation except on the first call to the function.

Reply:
There's `__builtin_popcountg()` for generic integers (note that `igraph_integer_t` may not be a `long long`).

Reply:
Annoyingly, `__builtin_popcountg()` is for the unreleased GCC 14 and Clang 19 only. So this is not an option.

Reply:
Luckily the functions are simple enough that the compilers (clang and gcc) are able to autovectorize given the correct compiler flags. Not sure about MSVC but won't spend any effort on it yet.

So now the only outstanding issue is handling when the operations are not available for MSVC.

Reply:
I added some checks in cmake for the MSVC implementation.